File: src/creaturetickbehavior/BugTickBehavior.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package creaturetickbehavior;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move.Moves;
import chipschallenge.tickbehaviors.BlockTickBehavior;

/**
 *
 * @author patrik
 */
public class BugTickBehavior implements BlockTickBehavior {

    private BugTickBehavior() {}
    private static BugTickBehavior mInstance = null;
    public static synchronized BugTickBehavior getInstance() {
        if(mInstance == null)
            mInstance = new BugTickBehavior();
        return mInstance;
    }

    @Override
    public void tick(Block caller) throws BlockContainerFullException {
        Moves m = caller.getFacing();
        outer: for(int i = 0; i < 2; i++) {
            switch(m) {
                case UP:
                    caller.setFacing(Moves.LEFT);
                    if(caller.move(Moves.LEFT))                    
                        break outer;
                case RIGHT:
                    caller.setFacing(Moves.UP);
                    if(caller.move(Moves.UP))                      
                        break outer;
                case DOWN:
                    caller.setFacing(Moves.RIGHT);
                    if(caller.move(Moves.RIGHT))                       
                        break outer;
                case LEFT:
                    caller.setFacing(Moves.DOWN);
                    if(caller.move(Moves.DOWN))                      
                        break outer;
                    else
                        m = Moves.UP;
            }
        }
    }

}

File: ChipsChallenge.Shared/Tickbehaviors/BugTickBehavior.cs
namespace ChipsChallenge.Shared.Tickbehaviors
{
    using Moves = Move.Moves;

    public class BugTickBehavior : IBlockTickBehavior
    {
        private BugTickBehavior()
        {
        }

        private static BugTickBehavior instance;

        public static BugTickBehavior Instance
        {
            get
            {
                lock (typeof(BugTickBehavior))
                {
                    return instance ?? (instance = new BugTickBehavior());
                }
            }
        }

        public virtual void Tick(Block caller)
        {
            Moves m = caller.Facing;
            
            switch (m)
            {
                case Moves.UP:
                    if (!caller.Move(Moves.LEFT))
                    {
                        if (!caller.Move(Moves.UP))
                        {
                            if (!caller.Move(Moves.RIGHT))
                            {
                                caller.Move(Moves.DOWN);
                            }
                        }
                    }
                    break;
                case Moves.LEFT:
                    if (!caller.Move(Moves.DOWN))
                    {
                        if (!caller.Move(Moves.LEFT))
                        {
                            if (!caller.Move(Moves.UP))
                            {
                                caller.Move(Moves.RIGHT);
                            }
                        }
                    }
                    break;
                case Moves.DOWN:
                    if (!caller.Move(Moves.RIGHT))
                    {
                        if (!caller.Move(Moves.DOWN))
                        {
                            if (!caller.Move(Moves.LEFT))
                            {
                                caller.Move(Moves.UP);
                            }
                        }
                    }
                    break;
                case Moves.RIGHT:
                    if (!caller.Move(Moves.UP))
                    {
                        if (!caller.Move(Moves.RIGHT))
                        {
                            if (!caller.Move(Moves.DOWN))
                            {
                                caller.Move(Moves.LEFT);
                            }
                        }
                    }
                    break;
            }
        }
    }
}